### PR TITLE
[dune] [api] Make some modules private.

### PR DESCRIPTION
**Pull request by ejgallego**
We mark some modules as an experiment wrt #8969 completion.

### CODE REVIEW

**Comment by ejgallego** [comment]
This was useless as originally submitted as plugins don't build vs the Dune-based build; I will rebase on top of #8729

**Comment by ejgallego** [comment]
Failures:
- `Class` used by Elpi and Mtac2.
- `Indschemes` used by Equations.
- `TQueue` by pidetop.
- `Declaredef` by tutorial.

**Comment by SkySkimmer** [comment]
Declaredef is definitely meant to be public.

**Comment by gares** [comment]
Isn't `Class` the module supposed to be used in order to get he list of coercions? Is there an higher level one I should use instead?

**Comment by ejgallego** [comment]
> Isn't `Class` the module supposed to be used in order to get he list of coercions? Is there an higher level one I should use instead?

I think so but I have not a clear idea myself.

**Comment by mattam82** [comment]
`Indschemes` should be public as well to let plugins ask for declaration of schemes they need.

File: INSTALL [hunk 1]
```diff
@@ -96,19 +93,14 @@ INSTALLATION PROCEDURE IN DETAILS (NORMAL USERS).
    bigger), you will also need the "ocamlopt" (or its native code version
    "ocamlopt.opt") command.
 
-2- Check that you have Camlp5 installed on your computer and that the
-   command "camlp5" lies in a directory which is present in your $PATH
-   environment variable path.  (You need Camlp5 in both bytecode and
-   native versions if your platform supports it).
-
-3- The uncompression and un-tarring of the distribution file gave birth
+2- The uncompression and un-tarring of the distribution file gave birth
```
Comment (jfehrle) [inline INSTALL]:
```suggestion
2- The uncompression and un-tarring of the distribution file created
```

File: INSTALL [hunk 2]
```diff
@@ -96,19 +93,14 @@ INSTALLATION PROCEDURE IN DETAILS (NORMAL USERS).
...
    to a directory named "coq-8.xx". You can rename this directory and put
```
Comment (jfehrle) [inline INSTALL]:
```suggestion
   a directory named "coq-8.xx". You can rename this directory and put
```

File: dev/ci/README.md [hunk 3]
```diff
@@ -84,10 +84,10 @@ unless these tests pass.
 
 We are currently running tests on the following platforms:
 
-- GitLab CI is the main CI platform. It tests the compilation of Coq, of the
-  documentation, and of CoqIDE on Linux with several versions of OCaml /
-  camlp5, and with warnings as errors; it runs the test-suite and tests the
-  compilation of several external developments.
+- GitLab CI is the main CI platform. It tests the compilation of Coq,
+  of the documentation, and of CoqIDE on Linux with several versions
```
Comment (jfehrle) [inline dev/ci/README.md]:
```suggestion
  the documentation, and CoqIDE on Linux with several versions
```

File: dev/doc/build-system.dune.md [hunk 4]
```diff
@@ -152,3 +152,31 @@ useful to Coq, some examples are:
...
+
+To teach Dune about the `.vo`, we use a small utility `coq_dune`,
+that will generate a `dune` file for each directory in `plugins` and
+`theories`. The code is pretty straightforward and declares build
+and install rules for each `.v` straight out of `coqdep`. Thus, our
+build strategy looks like this:
+
+1. Use `dune` to build `coqdep` and `coq_dune`.
+2. Use `coq_dune` to generate `dune` files for each directory with `.v` files.
+3. ?
```
Comment (jfehrle) [inline dev/doc/build-system.dune.md]: Can you elaborate on "?"?

File: tools/CoqMakefile.in [hunk 5]
```diff
@@ -34,13 +33,15 @@ COQLIBS           := $(COQMF_COQLIBS)
 COQLIBS_NOML      := $(COQMF_COQLIBS_NOML)
 CMDLINE_COQLIBS   := $(COQMF_CMDLINE_COQLIBS)
 LOCAL             := $(COQMF_LOCAL)
-COQLIB            := $(COQMF_COQLIB)
+# We add a workaround for #8731; if COQLIB is set externally, don't
```
Comment (jfehrle) [inline tools/CoqMakefile.in]:
```suggestion
# A workaround for #8731; if COQLIB is set externally, don't
```

**Review by jfehrle — APPROVED** [review]
Reviewed documentation--some wording suggestions.

**Comment by ejgallego** [comment]
Current list of private modules is more a demo of the feature than something we discussed deeply, all maintainers are invited to update the list of hidden modules if so they wish, for example for the kernel or pretyping.

**Comment by jfehrle** [comment]
My two cents:  Having the ability to mark things as private can be useful provided we're not overly rigid or overly aggressive about it.  User interface code tends to cut across abstractions in order to generate useful messages, support debugging by users and for other reasons.  The idea that you can decompose programs into a tidy set of modules breaks down as they grow larger.

**Comment by SkySkimmer** [comment]
CI isn't happy

**Comment by ppedrot** [comment]
This looks like a minor reenactment of the Coq API disaster to me. We handpick a few modules that we consider internal on murky grounds, check that the plugins we track do not depend on it (lo and behold: they do) and will eventually realize that some users we have never heard of were actually using the now private module in a critical way so we'll have to revert at least partially. I am not formally opposing the experiment (reducing the attack surface is a very good thing), but this does not seem like a effective protocol to define an API for external users to me.

**Comment by Zimmi48** [comment]
It sure is a less disastrous approach than the one that was used last time. And this allows to proceed cautiously, by marking very few modules private to start with (if possible by thinking through which modules should be marked private).

**Comment by ejgallego** [comment]
I agree with Pierre-Marie, indeed the PR is a bit arbitrary and it was done mostly to advertise the feature, in case it would be useful to us.

I think the advertising goal has been achieved [and indeed there are other interesting features of Dune that could be useful for Coq] , so indeed IMHO closing this would be fine too.

**Comment by Zimmi48** [comment]
So what is the actual status of the PR? Should it be closed or is this still proposed for merge? If yes, the actual reviewers should be the owners of the corresponding components, rather than the build system maintainers.

**Comment by ejgallego** [comment]
> So what is the actual status of the PR? Should it be closed or is this still proposed for merge? If yes, the actual reviewers should be the owners of the corresponding components, rather than the build system maintainers.

Indeed, thanks for the reminder, I was going to post that the PR is now working, so indeed we can either close, or if the maintainers like it, then we can merge. For vernac and toplevel, IMHO it is fine.

For @coq/stm-maintainers and @coq/tactics-maintainers we need their feedback.

**Review by ppedrot — APPROVED** [review]
The private modules from tactics/ definitely deserve to be private.

**Review by gares — COMMENTED** [review]
stm ok

**Review by Zimmi48 — APPROVED** [review]
OK then, let's merge. I propose to do so next Monday.

**Comment by Zimmi48** [comment]
@coqbot merge now

**Pull request merged**